// ftcommon/index.js
require('./lib/common/util');
require('./lib/common/console');
var fs = require('fs');
var umeng = require('./lib/thirdparty/umeng');
var runtime = null;
var debugMode = isDebugEnv();
function isDebugEnv(){
	if ( fs.existsSync('/etc/ftprod') )
		return false;
	return true;
}

function registerToRuntime(runtime){
	umeng.setRuntime(runtime);
	runtime.registerConst('debugMode', debugMode, {label:'调试模式'});
	runtime.register('umeng.debug', debugMode, null, {label:'调试推送', type:'bool'});
	runtime.register('umeng.android', true, null, {label:'推送Android设备', type:'bool'});
	runtime.register('umeng.ios', true, null, {label:'推送iOS设备', type:'bool'});
}

module.exports = {
	lang: require('./lib/common/lang'),
	date: require('./lib/common/date'),
	codes: require('./lib/common/codes'),
	result: require('./lib/common/result'),
	log: require('./lib/common/log'),
	logger: require('./lib/common/logger'),
	createSimpleRuntime: function(){
		var Runtime = require('./lib/common/runtime');
		runtime = new Runtime();
		registerToRuntime(runtime);
		runtime.start();
		return runtime;
	},
	createClusterRuntime: function(args){
		var Runtime = require('./lib/common/runtime');
		var RedisAdapter = require('./lib/common/RedisAdapter');
		var adapter = new RedisAdapter(args);
		runtime = new Runtime({adapter:adapter});
		registerToRuntime(runtime);
		runtime.start();
	},
	getRuntime: function(){
		return runtime;
	},
	Table: require('./lib/db/Table'),
	TableDao: require('./lib/db/TableDao'),
	mysql: require('./lib/db/Mysql'),
	Dao: require('./lib/db/Dao'),
	DaoDriver: require('./lib/db/DaoDriver'),

	alipay: require('./lib/thirdparty/alipay'),
	yeepay: require('./lib/thirdparty/yeepay'),
	wechat: require('./lib/thirdparty/wechat'),
	cloud: require('./lib/thirdparty/qiniu'),
	umeng: umeng,
	event: require('./lib/common/event'),

	isDebugEnv: function(){
		return debugMode;
	},

	useDebugConsole: function(){
		require('./lib/common/console');
	}
}
